compiler/src/dotty/tools/dotc/core/Types.scala

```diff
@@ -3448,6 +3451,47 @@ object Types {
       case that: OrType => tp1.eq(that.tp1) && tp2.eq(that.tp2) && isSoft == that.isSoft
       case _ => false
     }
+
+    /** Returns the set of non-union (leaf) types composing this union tree with Nothing types
+     *  absorbed by other types, if present. For example:<br>
+     *  `(A | B | C | B | (A & (B | C)) | Nothing)` returns `Set(A, B, C, (A & (B | C)))`.
+     */
+    private def gatherTreeUniqueMembersAbsorbingNothingTypes(using Context): Set[Type] = {
+      (tp1, tp2) match
+        case (l: OrType, r: OrType) =>
+          l.gatherTreeUniqueMembersAbsorbingNothingTypes ++ r.gatherTreeUniqueMembersAbsorbingNothingTypes
+        case (l: OrType, r) =>
+          if r.isNothingType
+          then l.gatherTreeUniqueMembersAbsorbingNothingTypes
+          else l.gatherTreeUniqueMembersAbsorbingNothingTypes + r
+        case (l, r: OrType) =>
+          if l.isNothingType
+          then r.gatherTreeUniqueMembersAbsorbingNothingTypes
+          else r.gatherTreeUniqueMembersAbsorbingNothingTypes + l
+        case (l, r) =>
+          if r.isNothingType then Set(l)
+          else if l.isNothingType then Set(r)
+          else Set(l, r)
+    }
+
+    /** Returns an equivalent union tree without repeated members and with Nothing types absorbed
+     *  by other types, if present. Weaker than LUB.
+     */
+    def deduplicatedAbsorbingNothingTypes(using Context): Type = {
```

Comment:
There are two problems with this approach

 - It looks inefficient. You need a fast mutable set based on identity not structural equality. Everything else would be needlessly slow.
 - Determinism. You need a set that has a reproducible order of elements between compiler runs. A LinkedHashSet would do, except that it's based on equality. A java.util.IdentityHashMap or a dotc.util.EqHashMap would be based on identity, but it's not deterministic. So maybe you have to write that set yourself. A combination of a ListBuffer for the elements and a EqHashMap could work, maybe.

Reply:
:heavy_check_mark: